ams/server/src/main/java/com/netease/arctic/server/optimizing/plan/CommonPartitionEvaluator.java

```diff
@@ -105,36 +108,66 @@ private boolean isDuplicateDelete(IcebergContentFile<?> delete) {
     return deleteExist;
   }
 
-  private void addFragmentFile(IcebergDataFile dataFile, List<IcebergContentFile<?>> deletes) {
+  private boolean addFragmentFile(IcebergDataFile dataFile, List<IcebergContentFile<?>> deletes) {
+    if (!fileShouldRewrite(dataFile, deletes)) {
+      return false;
+    }
     fragmentFileSize += dataFile.fileSizeInBytes();
     fragmentFileCount += 1;
 
     for (IcebergContentFile<?> delete : deletes) {
       addDelete(delete);
     }
+    return true;
   }
 
-  private void addSegmentFile(IcebergDataFile dataFile, List<IcebergContentFile<?>> deletes) {
-    segmentFileSize += dataFile.fileSizeInBytes();
-    segmentFileCount += 1;
-
-    if (shouldRewriteSegmentFile(dataFile, deletes)) {
+  private boolean addSegmentFile(IcebergDataFile dataFile, List<IcebergContentFile<?>> deletes) {
+    if (fileShouldRewrite(dataFile, deletes)) {
       rewriteSegmentFileSize += dataFile.fileSizeInBytes();
       rewriteSegmentFileCount += 1;
-    } else if (shouldRewritePosForSegmentFile(dataFile, deletes)) {
+    } else if (segmentFileShouldRewritePos(dataFile, deletes)) {
       rewritePosSegmentFileSize += dataFile.fileSizeInBytes();
       rewritePosSegmentFileCount += 1;
+    } else {
+      return false;
     }
+
+    segmentFileSize += dataFile.fileSizeInBytes();
+    segmentFileCount += 1;
     for (IcebergContentFile<?> delete : deletes) {
       addDelete(delete);
     }
+    return true;
   }
 
-  public boolean shouldRewriteSegmentFile(IcebergDataFile dataFile, List<IcebergContentFile<?>> deletes) {
+  protected boolean fileShouldFullOptimizing(IcebergDataFile dataFile, List<IcebergContentFile<?>> deleteFiles) {
+    if (config.isFullRewriteAllFiles()) {
+      return true;
+    }
+    if (isFragmentFile(dataFile)) {
+      return true;
+    }
+    // if a file is related any delete files or is not big enough, it should full optimizing
+    return !deleteFiles.isEmpty() || dataFile.fileSizeInBytes() < config.getTargetSize() * 0.9;
+  }
+
+  public boolean fileShouldRewrite(IcebergDataFile dataFile, List<IcebergContentFile<?>> deletes) {
+    if (isFullOptimizing()) {
+      return fileShouldFullOptimizing(dataFile, deletes);
+    }
+    if (isFragmentFile(dataFile)) {
```

Comment:
Sometimes there are some empty files(0 bytes), I think we should skip them, WDYT

Reply:
Self-optimizing has not yet considered how to deal with empty files, so we can open a new issue to discuss and address it.
My current view is that an empty file should be included as a fragment file for optimizing, to avoid this file lingering in the metadata indefinitely.

Reply:
OKay